Add changed field, metadata and index

njohner opened the pull request:
* Added a new `readonly` field `changed`, so that it cannot be modified by the users or through the API
* Added `changed` metadata and index
* `changed` is updated when an object is created (`IObjectCreatedEvent`) and generally when modified by a user (`IObjectModifiedEvent` except when the object is simply reindexed). The field is also updated for workflow status changes and when a document is checked-in or reverted to an older version.
* We add two upgrade steps: the first one creates the index (copy of modified date) and the empty metadata column, and the second one fills the metadata and reindexes the objects
* We show `changed` instead of `modified` in document listings. When the second upgrade has not been performed and solr is active, we do not show that column
* We add `changed` to the REST API summary.
* We upgrade `ftw.testing` as we need a bugfix allowing to get timezone aware now when freezing on a timezone naive datetime.

It seems that the `changed` index gets updated when the field is set, but I'm not sure why? I expected to have to explicitly reindex the object in the `_set_changed` method.

**Document listing without solr**
<img width="1250" alt="without_solr" src="https://user-images.githubusercontent.com/7374243/46950258-7743e980-d084-11e8-8709-a25a809746af.png">

**Document listing with solr but only with the first upgrade step. The `changed` column is not displayed**
<img width="1250" alt="with_solr" src="https://user-images.githubusercontent.com/7374243/46950259-7743e980-d084-11e8-9cfc-1ba5f71f48d9.png">

**Document listing with solr after both upgrade steps**
<img width="1250" alt="with_solr_after_upgrade" src="https://user-images.githubusercontent.com/7374243/46950260-7743e980-d084-11e8-8c9b-d27591cbe3a5.png">


resolves #4949

lukasgraf commented:
@njohner Tests are failing after your last push - but I think this is a flaky test failure unrelated to your changes, so I restarted the CI job.

lukasgraf commented:
@njohner Hmm. I just did a local test with running just the first upgrade step. After modification to a document's description, Catalog queries blow up in all kinds of nasty ways:

```python
2018-10-18 14:28:31 ERROR ZODB.Connection Couldn't load state for 0x29ef
Traceback (most recent call last):
  File "/Users/lukasgraf/Plone/eggs/ZODB3-3.10.7-py2.7-macosx-10.4-x86_64.egg/ZODB/Connection.py", line 860, in setstate
    self._setstate(obj)
  File "/Users/lukasgraf/Plone/eggs/ZODB3-3.10.7-py2.7-macosx-10.4-x86_64.egg/ZODB/Connection.py", line 914, in _setstate
    self._reader.setGhostState(obj, p)
  File "/Users/lukasgraf/Plone/eggs/ZODB3-3.10.7-py2.7-macosx-10.4-x86_64.egg/ZODB/serialize.py", line 612, in setGhostState
    state = self.getState(pickle)
  File "/Users/lukasgraf/Plone/eggs/ZODB3-3.10.7-py2.7-macosx-10.4-x86_64.egg/ZODB/serialize.py", line 605, in getState
    return unpickler.load()
  File "/Users/lukasgraf/Plone/eggs/pytz-2013b-py2.7.egg/pytz/__init__.py", line 291, in _p
    return unpickler(*args)
  File "/Users/lukasgraf/Plone/eggs/pytz-2013b-py2.7.egg/pytz/tzinfo.py", line 524, in unpickler
    tz = pytz.timezone(zone)
  File "/Users/lukasgraf/Plone/eggs/pytz-2013b-py2.7.egg/pytz/__init__.py", line 185, in timezone
    raise UnknownTimeZoneError(zone)
UnknownTimeZoneError: 'GMT+2'
2018-10-18 14:28:31 ERROR Zope.SiteErrorLog 1539865711.390.916973864773 http://localhost:8080/fd/ordnungssystem/fuehrung/gemeinderecht/dossier-13/tabbed_view/listing
Traceback (innermost last):
  Module ZPublisher.Publish, line 138, in publish
  Module ZPublisher.mapply, line 77, in mapply
  Module ZPublisher.Publish, line 48, in call_object
  Module ftw.tabbedview.browser.tabbed, line 164, in listing
  Module Products.Five.browser.metaconfigure, line 485, in __call__
  Module Products.Five.browser.pagetemplatefile, line 125, in __call__
  Module Products.Five.browser.pagetemplatefile, line 59, in __call__
  Module zope.pagetemplate.pagetemplate, line 132, in pt_render
  Module five.pt.engine, line 98, in __call__
  Module z3c.pt.pagetemplate, line 158, in render
  Module chameleon.zpt.template, line 297, in render
  Module chameleon.template, line 191, in render
  Module chameleon.template, line 171, in render
  Module 4de621f4a89083edb8d14314f531e440, line 184, in render
  Module five.pt.expressions, line 161, in __call__
  Module opengever.dossier.browser.overview, line 102, in is_subdossier_navigation_available
  Module opengever.dossier.base, line 174, in has_subdossiers
  Module opengever.dossier.base, line 201, in get_subdossiers
  Module Products.ZCatalog.Lazy, line 190, in __getitem__
  Module Products.ZCatalog.Catalog, line 122, in __getitem__
SystemError: error return without exception set
```

njohner commented on opengever/core/upgrades/20181009103059_fill_changed_date/upgrade.py:
```diff
@@ -0,0 +1,67 @@
+
+    def fill_changed_metadata(self):
+        """Set the 'changed' field on the object if it hasn't been set already"""
+        for obj in self.objects(query, 'Reindex proposal issuer'):
```
Ouch, the copy/paste fail

njohner commented on opengever/core/upgrades/20181009103059_fill_changed_date/upgrade.py:
```diff
@@ -0,0 +1,67 @@
+        catalog = self.getToolByName('portal_catalog')
+        changed_index = catalog._catalog.indexes["changed"]
+        changed_index.clear()
+        self.catalog_reindex_objects(query, idxs=["changed", "object_provides"])
```
Yes in the upgrade I know the `reindexObject` updates the metadata. It's more in the event handlers, I do not explicitely reindex anything and both the metadata and index seem to get updated (according to the tests).

njohner commented on opengever/base/indexes.py:
```diff
@@ -35,3 +38,11 @@ def bundle_guid_indexer(obj):
+def changed_indexer(obj):
+    if IOpenGeverChangedMarker.providedBy(obj):
+        # Entry in the Catalog should be a zope DateTime
+        return DateTime(IOpenGeverChanged(obj).changed)
```
I'm not actually sure that this is necessary. This leads to the metadata being a Zope DateTime I think (see `test_indexers.py`).
Other date indexers we have do not do that: https://github.com/4teamwork/opengever.core/blob/master/opengever/document/indexers.py#L108-L115

njohner replied:
The index itself is anyway stored as an integer, and the `DateIndex` class takes care of transforming whatever is input (Zope DateTime, datetime, string) to its internal format.

njohner commented:
@deiferni 
I think you inherited the review of this one 😓. Enjoy...

njohner commented:
@deiferni 
I think you inherited the review of this one 😓. Enjoy...

njohner commented:
@deiferni 
I think you inherited the review of this one 😓. Enjoy...

deiferni commented on opengever/tabbedview/browser/tabs.py:
```diff
@@ -189,6 +191,14 @@ class Documents(BaseCatalogListingTab):
 
+    @property
+    def columns(self):
+        if (api.portal.get_registry_record('use_solr', interface=ISearchSettings)
```
i would take this one step further and already create an issue (if not done already), and link to the issue in your comments. that way we have a chance the `XXX` comments will be processed eventually 😉.

deiferni commented on opengever/base/behaviors/changed.py:
```diff
@@ -0,0 +1,62 @@
+        # Make sure we have a python datetime
+        assert isinstance(value, datetime), "changed should be a python datetime"
+        # Make sure we have a timezone naive datetime
+        assert value.tzinfo is None, "changed should be timezone naive"
```
as discussed we should try again with timezone aware datetimes as the original problem ist not so clear any more 😉

njohner commented:
I think this is ready

njohner commented on opengever/core/upgrades/20181008173501_add_changed_date/upgrade.py:
```diff
@@ -0,0 +1,46 @@
+        self.set_changed_flag()
+
+    def add_changed_index(self):
+        """ We fill the index by copying the data from the 'modified' index,
```
I only partially remember why we need to have the index set even if the second upgrade step has not been executed. But for sure this allows tables to work before having the second upgrade, and also allows queries. And it's inexpensive to do.

deiferni replied:
there is only a benefit for querying, but i don't think it is used in a query. it won't show up in tables as the metadata is not updated yet.
so if there is no need to do this i would tend to drop it as there is a certain risk when messing with catalog internals.

njohner replied:
It will show up in the tables, if solr is not used (hence the fallback for the metadata)

deiferni replied:
as just discussed, there may be trouble when the index is empty, despite the fallback to modified on display the index will be used for sorting. so 👍.

njohner commented:
Ok I went over this once more, to make sure I had included @deiferni 's comments.
I just corrected a comment:
```
git diff origin/nj_modified
diff --git a/opengever/base/behaviors/changed.py b/opengever/base/behaviors/changed.py
index e78777dd3..bd0bed92c 100644
--- a/opengever/base/behaviors/changed.py
+++ b/opengever/base/behaviors/changed.py
@@ -44,7 +44,7 @@ class Changed(object):
         return self.context.changed
 
     def _set_changed(self, value):
-        # Make sure we have a python datetime
+        # Make sure we have a zope datetime
         if isinstance(value, datetime):
             value = DateTime(value)
         assert isinstance(value, DateTime), "changed should be a zope datetime"
```
apart from that I did not find anything else. So on my side I think this is ready to ship.

phgross commented:
@njohner tested with and without solr, seems to work fine 👍 --> 🚢
